DxDispatch/src/model/OnnxParsers.cpp

```diff
@@ -0,0 +1,173 @@
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64:       return true;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE:      return true;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX64:   return false;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX128:  return false;
+    default: return false;
+    }
+}
+
+DML_TENSOR_DATA_TYPE OnnxParsers::ConvertOnnxTensorDataType(ONNXTensorElementDataType dataType)
+{
+    switch (dataType)
+    {
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8:   return DML_TENSOR_DATA_TYPE_UINT8;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8:    return DML_TENSOR_DATA_TYPE_INT8;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16:  return DML_TENSOR_DATA_TYPE_UINT16;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16:   return DML_TENSOR_DATA_TYPE_INT16;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16: return DML_TENSOR_DATA_TYPE_FLOAT16;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32:   return DML_TENSOR_DATA_TYPE_INT32;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32:  return DML_TENSOR_DATA_TYPE_UINT32;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT:   return DML_TENSOR_DATA_TYPE_FLOAT32;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64:  return DML_TENSOR_DATA_TYPE_UINT64;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64:   return DML_TENSOR_DATA_TYPE_INT64;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE:  return DML_TENSOR_DATA_TYPE_FLOAT64;
+    default: throw std::invalid_argument("Unsupported tensor type");
+    }
+}
+
+static uint32_t OnnxTensorDataTypeSize(ONNXTensorElementDataType dataType)
+{
+    switch (dataType)
+    {
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8:   return 1;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8:    return 1;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16:  return 2;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16:   return 2;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16: return 2;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32:   return 4;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32:  return 4;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT:   return 4;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64:  return 8;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64:   return 8;
+    case ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE:  return 8;
+    default: throw std::invalid_argument("Unsupported tensor type");
+    }
+}
+
+Model OnnxParsers::ParseModel(const std::filesystem::path& filePath, gsl::span<const std::pair<std::string, uint32_t>> freeDimOverrides)
+{
+    const OrtApi& ortApi = Ort::GetApi();
+
+    Ort::SessionOptions sessionOptions;
+    sessionOptions.SetExecutionMode(ExecutionMode::ORT_SEQUENTIAL);
+    sessionOptions.DisableMemPattern();
+    sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
+
+    for (auto& freeDimOverride : freeDimOverrides)
+    {
+        ortApi.AddFreeDimensionOverrideByName(sessionOptions, freeDimOverride.first.c_str(), freeDimOverride.second);
+    }
+
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_DML(sessionOptions, 0));
```

Comment:
>OrtSessionOptionsAppendExecutionProvider_DML [](http://example.com/codeflow?start=22&length=44)

Note this API is deprecated. I know it still works, but and the newer one is more verbose than calling the global free function, but if others copy this code as a reference, then it would be nice for them to see the new one:

```
    const OrtDmlApi* ortDmlApi;
    THROW_IF_NOT_OK(Ort::GetApi().GetExecutionProviderApi("DML", ORT_API_VERSION, reinterpret_cast<const void**>(&ortDmlApi)));
    ortDmlApi->SessionOptionsAppendExecutionProvider_DML(sessionOptions, 0);
```

This has the v-table advantage that it's dynamically callable depending on support rather than a hard build/linker error when binding to onnxruntime.dll, depending on what it's built with.